File: frontend/src/types/message.ts
import { Channel } from './channel';
import { MessageUser, User } from './user';

export class Message {
	id?: number;
	userId: number;
	user: User | MessageUser;
	text: string;
	time: number;
	channelId: number;
	channel: Channel;

	constructor(
		user: User | MessageUser,
		text: string,
		time: number = Date.now(),
		channel: Channel,
		id?: number
	) {
		this.user = user;
		this.userId = user.id;
		this.text = text;
		this.time = time;
		this.channel = channel;
		this.channelId = channel.id;
		if (id) this.id = id;
	}

	getSerialized() {
		return {
			userId: this.userId,
			channelId: this.channelId,
			text: this.text,
			time: this.time
		};
	}
}

File: backend/Server/Models/Message.cs
using RealTimeChatServer.Websockets.Payloads;

namespace RealTimeChatServer.Models;

public partial class Message
{
	public int Id { get; set; }
	public int UserId { get; set; }
	public int ChannelId { get; set; }
	public string Text { get; set; } = null!;
	public long Time { get; set; }

	// Navigational Properties
	public virtual User User { get; set; } = null!;
	public virtual Channel Channel { get; set; } = null!;

	public WsMessageDto ToDTO()
	{
		return new WsMessageDto
        {
			Id = Id,
			Text = Text,
			Time = Time,
			UserId = UserId,
			ChannelId = ChannelId,
			User = User?.ToDTO(),
			Channel = Channel?.ToDTO()
		};
	}

	public WsMessageDto ToMinimalDTO()
	{
		return new WsMessageDto
        {
			Id = Id,
			Text = Text,
			Time = Time,
			UserId = UserId,
			ChannelId = ChannelId
		};
	}
}
